File: packages/blendlink/src/webRuntimeLimits.ts
import { existsSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

/**
 * The pinned web runtime's hard ceilings, authored once as data beside the
 * Blender addon and read here so the compiled-artifact gate and the Blender
 * preflight cannot disagree. `web_runtime_limits.py` validates the same file
 * with the same rules; a headless check asserts both readings match, which is
 * the only thing that keeps two languages honest about one number.
 *
 * These are not budgets or recommendations. Crossing one of them produces a
 * black character or an invisible mesh with no runtime error at all, which is
 * why each record carries the symptom and the artist action rather than only
 * the number.
 */
export interface WebRuntimeLimit {
  id: string
  measure: string
  maximum: number
  summary: string
  symptom: string
  action: string
  evidence: string[]
}

export interface WebRuntimeLimitRegistry {
  schemaVersion: 1
  policy: string
  runtime: string
  limits: WebRuntimeLimit[]
}

const here = dirname(fileURLToPath(import.meta.url))

function registryPath(): string {
  const candidates = [
    // Packaged: copy-assets ships the addon beside the compiled JS.
    join(here, 'addon', 'web_runtime_limits.json'),
    // Workspace: running from source.
    join(here, '..', '..', 'blender-addon', 'web_runtime_limits.json'),
  ]
  const found = candidates.find((candidate) => existsSync(candidate))
  if (!found) {
    throw new Error(
      'Blendlink cannot read its web runtime limits. Looked in: ' +
        candidates.join(', ') +
        '. This file carries the ceilings that make skinning and UV binding ' +
        'refusals possible; without it Blendlink would publish scenes it ' +
        'knows cannot render.',
    )
  }
  return found
}

function validate(data: unknown): WebRuntimeLimitRegistry {
  const registry = data as WebRuntimeLimitRegistry
  if (!registry || typeof registry !== 'object' || registry.schemaVersion !== 1) {
    throw new Error('web runtime limit registry schemaVersion must be 1')
  }
  if (!Array.isArray(registry.limits) || registry.limits.length === 0) {
    throw new Error('web runtime limit registry limits must be a nonempty array')
  }
  const seen = new Set<string>()
  for (const [index, limit] of registry.limits.entries()) {
    if (!limit || typeof limit.id !== 'string' || !limit.id.trim()) {
      throw new Error(`limits[${index}].id must be non-empty text`)
    }
    if (seen.has(limit.id)) {
      throw new Error(`duplicate web runtime limit id ${limit.id}`)
    }
    seen.add(limit.id)
    if (!Number.isInteger(limit.maximum) || limit.maximum <= 0) {
      throw new Error(`limits[${index}].maximum must be a positive integer`)
    }
    for (const field of ['measure', 'summary', 'symptom', 'action'] as const) {
      if (typeof limit[field] !== 'string' || !limit[field].trim()) {
        throw new Error(`limits[${index}].${field} must be non-empty text`)
      }
    }
    if (!Array.isArray(limit.evidence) || limit.evidence.length === 0) {
      throw new Error(
        `limits[${index}].evidence must be a nonempty array; a ceiling ` +
          'without evidence is a guess',
      )
    }
  }
  return registry
}

let cache: WebRuntimeLimitRegistry | undefined

export function webRuntimeLimits(): WebRuntimeLimitRegistry {
  if (!cache) {
    cache = validate(JSON.parse(readFileSync(registryPath(), 'utf8')))
  }
  return cache
}

export function webRuntimeLimit(id: string): WebRuntimeLimit {
  const registry = webRuntimeLimits()
  const found = registry.limits.find((limit) => limit.id === id)
  if (!found) {
    throw new Error(
      `unknown web runtime limit ${id}; the registry declares ` +
        registry.limits.map((limit) => limit.id).sort().join(', '),
    )
  }
  return found
}

/** The artist-facing sentence pair every refusal on this limit reuses. */
export function webRuntimeLimitConsequence(id: string): string {
  const limit = webRuntimeLimit(id)
  return `${limit.symptom} ${limit.action}`
}

File: packages/blender-addon/web_runtime_limits.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Hard ceilings of the pinned web runtime, shared with the TypeScript side.

The numbers in ``web_runtime_limits.json`` are the reason several refusals
exist, and they have to be identical in two languages: Blender refuses before
a two-hour bake, and the compiled-artifact gate refuses again on the finished
GLB because the exporter is not the only thing that can put a joint or a UV
set into it. A constant copied into both languages drifts silently, so the
number is authored once as data and both readers assert against it.

Empty is not a valid state here: a limit is removed only when the pinned
runtime stops having it.
"""
from __future__ import annotations

import json
from pathlib import Path

REGISTRY_PATH = Path(__file__).with_name("web_runtime_limits.json")

_registry_cache = None


def _text(value, field):
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field} must be non-empty text")
    return value.strip()


def validate_registry(data):
    """Fail loudly on a malformed registry rather than defaulting a ceiling."""
    if not isinstance(data, dict) or data.get("schemaVersion") != 1:
        raise ValueError("web runtime limit registry schemaVersion must be 1")
    _text(data.get("policy"), "web runtime limit registry policy")
    _text(data.get("runtime"), "web runtime limit registry runtime")
    limits = data.get("limits")
    if not isinstance(limits, list) or not limits:
        raise ValueError(
            "web runtime limit registry limits must be a nonempty array"
        )
    seen = set()
    resolved = {}
    for index, limit in enumerate(limits):
        if not isinstance(limit, dict):
            raise ValueError(f"limits[{index}] must be an object")
        limit_id = _text(limit.get("id"), f"limits[{index}].id")
        if limit_id in seen:
            raise ValueError(f"duplicate web runtime limit id {limit_id}")
        seen.add(limit_id)
        maximum = limit.get("maximum")
        if not isinstance(maximum, int) or isinstance(maximum, bool) \
                or maximum <= 0:
            raise ValueError(
                f"limits[{index}].maximum must be a positive integer"
            )
        for field in ("measure", "summary", "symptom", "action"):
            _text(limit.get(field), f"limits[{index}].{field}")
        evidence = limit.get("evidence")
        if not isinstance(evidence, list) or not evidence:
            raise ValueError(
                f"limits[{index}].evidence must be a nonempty array; a ceiling "
                "without evidence is a guess"
            )
        for position, item in enumerate(evidence):
            _text(item, f"limits[{index}].evidence[{position}]")
        resolved[limit_id] = limit
    return resolved


def registry():
    global _registry_cache
    if _registry_cache is None:
        try:
            data = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
        except (OSError, ValueError) as error:
            raise ValueError(
                f"cannot read web runtime limits at {REGISTRY_PATH}: {error}"
            ) from error
        _registry_cache = validate_registry(data)
    return _registry_cache


def limit(limit_id: str) -> dict:
    """One validated limit record, or a loud failure naming the known ids."""
    found = registry().get(limit_id)
    if found is None:
        raise KeyError(
            f"unknown web runtime limit {limit_id!r}; the registry declares "
            + ", ".join(sorted(registry()))
        )
    return found


def maximum(limit_id: str) -> int:
    return int(limit(limit_id)["maximum"])


def consequence(limit_id: str) -> str:
    """The artist-facing sentence pair every refusal on this limit reuses."""
    record = limit(limit_id)
    return f"{record['symptom']} {record['action']}"
